### PR TITLE
Relation operations

### PR DESCRIPTION
**Pull request by Meli-0xFF**
Added functions from assignment, refactored.

### CODE REVIEW

File: relation.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    def reflexive_transitive_closure(self):
+        closure = self.R.union(set(map(lambda m : (m, m), self.M)))
+        while True:
+            closure_until_now = closure | set((x,w) for x,y in closure for q,w in closure if q == y)
+            if closure_until_now == closure: break
+            closure = closure_until_now
+        return Relation(self.M, closure_until_now)
+
+def get_relation_class(M):
+    return Relation(M)
```
Comment (krtko1) [inline relation.py]: careful if you would use `PClass` all parameters will be positional, so you will need use `Relation(M=M)`

File: relation.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        return set() == set(filter(lambda m : True if not self.contains(m, m) else False, self.M))
+
+    def is_symetric(self):
+        return set() == set(filter(lambda t : True if not self.contains(t[1], t[0]) else False, self.R))
+
+    def is_transitive(self):
+        return set() == set((t1, t2) for t1 in self.R for t2 in self.R if t1[1] == t2[0] and not self.contains(t1[0], t2[1]))
+
+    def reflexive_transitive_closure(self):
+        closure = self.R.union(set(map(lambda m : (m, m), self.M)))
```
Comment (krtko1) [inline relation.py]: It would be nice if you always use same syntax, choose one `union` or `|`.

File: relation.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        return Relation(self.M, R=set((t1[0], t2[1]) for t1 in self.R for t2 in R2.R if t1[1] == t2[0]))
+
+    def is_reflexive(self):
+        return set() == set(filter(lambda m : True if not self.contains(m, m) else False, self.M))
+
+    def is_symetric(self):
+        return set() == set(filter(lambda t : True if not self.contains(t[1], t[0]) else False, self.R))
+
+    def is_transitive(self):
+        return set() == set((t1, t2) for t1 in self.R for t2 in self.R if t1[1] == t2[0] and not self.contains(t1[0], t2[1]))
```
Comment (krtko1) [inline relation.py]: This line is quite long but it's not huge problem

File: relation.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        return Relation (self.M, R=self.R & R2.R)
+
+    def subtraction(self, R2):
+        return Relation(self.M, R=self.R - R2.R)
+
+    def inverse(self):
+        return Relation(self.M, R=set(map(lambda t : t[::-1], self.R)))
+ 
+    def composition(self, R2):
+        return Relation(self.M, R=set((t1[0], t2[1]) for t1 in self.R for t2 in R2.R if t1[1] == t2[0]))
```
Comment (krtko1) [inline relation.py]: You could use `{(t1[0], t2[1]) for t1 in self.R for t2 in R2.R if t1[1] == t2[0]}`, but it's matter of preferences.

File: relation.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+class Relation:
+    def __init__(self, M, R=set()):
```
Comment (krtko1) [inline relation.py]: then don't use `__init__` but define `M` and `R`, though `field` like in this example https://pyrsistent.readthedocs.io/en/latest/intro.html#pclass

File: relation.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+class Relation:
```
Comment (krtko1) [inline relation.py]: If you want better performance you should extend `PCass` from `pyrsistent`

Comment (Meli-0xFF) [reply relation.py]: Done, you can check.

File: relation.py (hunk 5 above)
Comment (Meli-0xFF) [reply relation.py]: __init__ removed, now using fields.

File: relation.py (hunk 4 above)
Comment (Meli-0xFF) [reply relation.py]: I like it more with pset() (set() changed to pset())

File: relation.py (hunk 3 above)
Comment (Meli-0xFF) [reply relation.py]: I like it in one line

File: relation.py (hunk 2 above)
Comment (Meli-0xFF) [reply relation.py]: repaired, now using |

File: relation.py (hunk 1 above)
Comment (Meli-0xFF) [reply relation.py]: Thank you, did it.

File: relation.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        return self.R.__iter__()
+
+    def __next__(self):
+        return self.R.__next__()
+
+    def contains(self, x, y):
+        return (x,y) in self.R
+
+    def add(self, x, y):
+        return Relation(M=self.M, R=self.R | pset((x,y)))
```
Comment (krtko1) [inline relation.py]: I think you do not add `x` and `y` correctly

File: relation.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    def remove(self, x, y):
+        return Relation(M=self.M, R=self.R - pset((x,y)))
```
Comment (krtko1) [inline relation.py]: I think you do not remove `x` and `y` correctly

File: relation.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        return Relation (M=self.M, R=self.R & R2.R)
+
+    def subtraction(self, R2):
+        return Relation(M=self.M, R=self.R - R2.R)
+
+    def inverse(self):
+        return Relation(M=self.M, R=pset(map(lambda t : t[::-1], self.R)))
+ 
+    def composition(self, R2):
+        return Relation(self.M, R=pset((t1[0], t2[1]) for t1 in self.R for t2 in R2.R if t1[1] == t2[0]))
```
Comment (krtko1) [inline relation.py]: you forgot that now is `M` positional argument

File: relation.py (hunk 2 above)
Comment (krtko1) [reply relation.py]: Nice :)

File: relation.py (hunk 9 above)
Comment (Meli-0xFF) [reply relation.py]: Done.

File: relation.py (hunk 8 above)
Comment (Meli-0xFF) [reply relation.py]: Changed to correct format.

File: relation.py (hunk 7 above)
Comment (Meli-0xFF) [reply relation.py]: Changed to correct format.

**Pull request merged**